### ISSUE TITLE
[Ramda] `…rest: Array<void>` allows undefined to be passed, wrong function definition gets chosen

### ISSUE THREAD

**Issue by hokennethk** [flowtype/flow-typed#1965]
Hey everyone! I have both an issue and question.

tldr: passing `undefined` as function arg causes flow to choose the "curried" type definition for that function (i.e. in `equals`)

Consider the following (see [flow repl](https://flow.org/try/#0CYUwxgNghgTiAEAzArgOzAFwJYHtXxAEdkoIBnAHgBUA+ACgA8AueKgGngDpu4yMWAgjBhQAnhQBuOLMBoBKFnVEsqc+AF4a8AEY4cEEFFQBuAFChIsBCnTY8BYqUq1GKjstYKdeg0bOmMUQAHBAAFGBwgsg14AG9TeHgwAAssCGA4fBYAfiERcQB5bQArcAwaUwBfBPhTG0xcfEQ9eDooeBZwyLIObQ74Lqi1Fl19Q3x4gEg4DGQYfCIScjbOFLSMkFRe1dT0zLkq0yA)):

```js
// current type definition in flow-typed project
declare function equals<T>(x: T, ...rest: Array<void>): (y: T) => boolean;
declare function equals<T>(x: T, y: T): boolean;

type Props = {
  children : ?Array<Object>
}
  
function foo (a : Props, b : Props) : boolean {
	return equals(a.children, b.children)
}
```

This results in the following error:
```sh
9: 	return equals(a.children, b.children)           
      ^ Cannot return `equals(...)` because function type [1] is incompatible with boolean [2].
  References:
  1: declare function equals<T>(x: T, ...rest: Array<void>): (y: T) => boolean;
                                                             ^ [1]
  8: function foo (a : Props, b : Props) : boolean {
                                           ^ [2]
```

It looks like `return equals(a.children, b.children)` matches the first definition of equals (`function equals<T>(x: T, ...rest: Array<void>): (y: T) => boolean;`) when it should be matching the second, since both arguments are passed. My guess is that since the `children` property of `type Props` has a `?`, it can be considered `null` or `void`, which matches up with `...rest: Array<void>`

This also fails the cli `run-tests` command if the following is added to the ramda tests:
`const eqVoid: boolean = _.equals(undefined, undefined)`

So I have a few questions/comments up for discussion:
1) This seems like unexpected behavior in terms of how we expect ramda `equals` to behave
2) I'm no flow expert so I'm hoping to get some insight into what the `...rest: Array<void>` syntax provides? It's used in many of the types here. Why not do something like this instead: `declare function equals<T>(x: T): (y: T) => boolean;`?
3) Alternatively, should we consider using `...rest: Array<empty>` instead? This way, we won't match up with `undefined`/`void` when that is passed as an argument. According to [this PR](https://github.com/facebook/flow/issues/4090), the `empty` type is:
  * > a “type that is not satisfied by any value”

Further Notes
* suggestions #2 and #3 above fix the `const eqVoid: boolean = _.equals(undefined, undefined)` test suggested. #2 breaks the codebase I'm working on (for some reason, still investigating), but #3 works :)

**Comment by AndrewSouthpaw** [comment]
cc @LoganBarnett 

We're slowly redefining that pattern, which you're quite correct has some limitations. Here's the current approach, demonstrated through `#reduce`

```javascript
  declare type Reduce = (<A, B>(
    fn: (acc: A, elm: B) => A
  ) => ((init: A) => (xs: Array<B> | $ReadOnlyArray<B>) => A) &
    ((init: A, xs: Array<B> | $ReadOnlyArray<B>) => A)) &
    (<A, B>(
      fn: (acc: A, elm: B) => A,
      init: A
    ) => (xs: Array<B> | $ReadOnlyArray<B>) => A) &
    (<A, B>(
      fn: (acc: A, elm: B) => A,
      init: A,
      xs: Array<B> | $ReadOnlyArray<B>
    ) => A);

  declare var reduce: Reduce;
```

Should be easy enough to adapt for `equals`, but let me know if you have questions. Feel free to do a few, try it on for size, and submit a PR. We're all about incremental improvements.
